### DISCUSSION TITLE
Why does install with CSI plugin work differently now (1.9?)

### DISCUSSION THREAD

**Discussion by scaleoutsean** [vmware-tanzu/velero#5415] [Community support Q&A] [answered]
- K8s v1.23.5
- V 1.9.2

I have 2 questions on the changes in Velero CSI behavior during installation.

First example. This used to work (with a different CSI plugin version and Velero 1.5):

```
velero install --features=EnableCSI
```

Now it results in an error:

```
An error occurred: --provider is required
```

Why do we need a provider here? Maybe all I want is local CSI snapshots?

It seems CSI plugin cannot be specified if AWS plugin is provided:

```
velero install \
  --features=EnableCSI \
  --provider aws \
  --bucket velero \
  --secret-file ./credentials-velero \
  --use-volume-snapshots=false \
  --backup-location-config region=us-east-1,s3ForcePathStyle="true",s3Url=http://2.2.2.2 \
  --plugins velero/velero-plugin-for-aws:v1.5.1 \
  # --plugins velero/velero-plugin-for-csi:v0.3.1
```

If the last line is not marked-out, we get this error:

```
time="2022-09-30T11:20:10Z" level=error msg="Current BackupStorageLocations available/unavailable/unknown: 0/0/1, BackupStorageLocation \"default\" is unavailable: unable to locate ObjectStore plugin named velero.io/aws)" controller=backup-storage-location logSource="pkg/controller/backup_storage_location_controller.go:171"
time="2022-09-30T11:20:10Z" level=error msg="Error getting a backup store" backup-storage-location=velero/default controller=backup-storage-location error="unable to locate ObjectStore plugin named velero.io/aws" logSource="pkg/controller/backup_storage_location_controller.go:127"
```

Is it no longer possible to specify which version of CSI plugin to use?

**Comment by sseago** [comment, marked as answer]
A provider is required for backup storage. Even CSI backups need an ObjectStore plugin to store the k8s resources in the backup.
Regarding the plugins, I suspect that the `--plugins` arg is taking only the last one rather than combining separate `--plugins` args. The usualy way to do this would be `--plugins velero/velero-plugin-for-aws:v1.5.1,velero/velero-plugin-for-csi:v0.3.1`

**Reply by scaleoutsean** [reply]
`--plugins velero/velero-plugin-for-aws:v1.5.1,velero/velero-plugin-for-csi:v0.3.1` does work, thank you. I wish that was documented somewhere, since it's so common (i.e. you can't use CSI without using a provider, for which we need a plugin, i.e. 2 plugins are required).
In v1.5.2 I remember provider wasn't required - I think that may have changed since then, unless I had forgotten. I think it'd be better if a provider for backup wasn't *required*. People who just want to manage snapshots could use Velero for that and nothing else. But if that's how it works now, that's fine.

**Reply by sseago** [reply]
Hmm. It could be that validation changed, but it's never been possible to perform a backup without one, since there would be no place to store it.

**Reply by sseago** [reply]
CSI plugin requires backing up k8s resources as part of it, which requires a provider, even if all you want is volumes.

**Reply by scaleoutsean** [reply]
"Local backups" described [here](https://docs.portworx.com/operations/operate-kubernetes/storage-operations/disaster-recovery/#local-backups) is an example of a use case like that with Velero, I think. While having a CSI snapshot doesn't mean you have a backup (i.e. copy of your data independent from the original), if it is available and can be restored, that can be the first line of defense.

**Reply by sseago** [reply]
I'm seeing "velero backup create" in that example, which means you are creating a backup. Velero can't create a backup without a provider, since there's no place to store the k8s resources, and even restoring PV data requires the PV/PVC k8s resources in the backup (as well as VolumeSnapshot, etc. for CSI backups).
